packages/amplify-graphql-index-transformer/src/resolvers.ts

```diff
@@ -272,3 +298,217 @@ function setQuerySnippet(config: PrimaryKeyDirectiveConfiguration, ctx: Transfor
 
   return block(`Set query expression for key`, expressions);
 }
+
+export function appendSecondaryIndex(config: IndexDirectiveConfiguration, ctx: TransformerContextProvider): void {
+  const { name, object, primaryKeyField } = config;
+  const table = getTable(ctx, object) as any;
+  const keySchema = getDdbKeySchema(config);
+  const attrDefs = attributeDefinitions(config, ctx);
+  const primaryKeyPartitionKeyName = primaryKeyField?.name?.value ?? 'id';
+  const partitionKeyName = keySchema[0]?.attributeName;
+  const sortKeyName = keySchema?.[1]?.attributeName;
+  const partitionKeyType = attrDefs.find(attr => attr.attributeName === partitionKeyName)?.attributeType ?? 'S';
+  const sortKeyType = sortKeyName ? attrDefs.find(attr => attr.attributeName === sortKeyName)?.attributeType ?? 'S' : undefined;
+  const defaultGSI = ctx.featureFlags.getBoolean('secondaryKeyAsGSI', false);
+
+  if (!defaultGSI && primaryKeyPartitionKeyName === partitionKeyName) {
+    // Create an LSI.
+    table.addLocalSecondaryIndex({
+      indexName: name,
+      projectionType: 'ALL',
+      sortKey: sortKeyName
+        ? {
+            name: sortKeyName,
+            type: sortKeyType,
+          }
+        : undefined,
+    });
+  } else {
+    // Create a GSI.
+    table.addGlobalSecondaryIndex({
+      indexName: name,
+      projectionType: 'ALL',
+      partitionKey: {
+        name: partitionKeyName,
+        type: partitionKeyType,
+      },
+      sortKey: sortKeyName
+        ? {
+            name: sortKeyName,
+            type: sortKeyType,
+          }
+        : undefined,
+      readCapacity: cdk.Fn.ref(ResourceConstants.PARAMETERS.DynamoDBModelTableReadIOPS),
+      writeCapacity: cdk.Fn.ref(ResourceConstants.PARAMETERS.DynamoDBModelTableWriteIOPS),
+    });
+
+    // At the L2 level, the CDK does not handle the way Amplify sets GSI read and write capacity
+    // very well. At the L1 level, the CDK does not create the correct IAM policy for accessing the
+    // GSI. To get around these issues, keep the L1 and L2 GSI list in sync.
+    const cfnTable = table.table;
+    cfnTable.globalSecondaryIndexes = appendIndex(cfnTable.globalSecondaryIndexes, {
+      indexName: name,
+      keySchema,
+      projection: { projectionType: 'ALL' },
+      provisionedThroughput: cdk.Fn.conditionIf(ResourceConstants.CONDITIONS.ShouldUsePayPerRequestBilling, cdk.Fn.ref('AWS::NoValue'), {
+        ReadCapacityUnits: cdk.Fn.ref(ResourceConstants.PARAMETERS.DynamoDBModelTableReadIOPS),
+        WriteCapacityUnits: cdk.Fn.ref(ResourceConstants.PARAMETERS.DynamoDBModelTableWriteIOPS),
+      }),
+    });
+  }
+}
+
+function appendIndex(list: any, newIndex: any): any[] {
```

Comment:
nit: consider using `$TSAny` instead of `any` if a type cannot be defined

Reply:
I didn't use `$TSAny` to try avoiding taking on a dependency on the CLI core since the Transformer can be used without the CLI.

Reply:
makes sense, thanks